src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,309 @@
+<?php
+/*
+ * Copyright 2023-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Model;
+
+use AppendIterator;
+use ArrayAccess;
+use ArrayIterator;
+use CallbackFilterIterator;
+use Countable;
+use IteratorAggregate;
+use JsonSerializable;
+use MongoDB\BSON\PackedArray;
+use MongoDB\Codec\CodecLibrary;
+use MongoDB\Codec\LazyBSONCodecLibrary;
+use MongoDB\Exception\InvalidArgumentException;
+use ReturnTypeWillChange;
+
+use function array_filter;
+use function array_key_exists;
+use function array_keys;
+use function array_map;
+use function array_values;
+use function count;
+use function is_array;
+use function is_numeric;
+use function iterator_to_array;
+use function max;
+use function MongoDB\recursive_copy;
+use function sprintf;
+use function trigger_error;
+
+use const E_USER_WARNING;
+
+/**
+ * Model class for a BSON array.
+ *
+ * The internal data will be filtered through array_values() during BSON
+ * serialization to ensure that it becomes a BSON array.
+ *
+ * @template TValue
+ * @template-implements ArrayAccess<int, TValue>
+ * @template-implements IteratorAggregate<int, TValue>
+ */
+final class LazyBSONArray implements ArrayAccess, Countable, IteratorAggregate, JsonSerializable
```

Comment:
I've considered implementing `Serializable` for convenience. We can certainly consider this for these classes. I would not implement `Unserializable`, as that would defeat the purpose of the classes being lazy (as they'd receive decoded structures already).

Reply:
I think Serializable makes sense since it might help users avoid a hard-to-debug mistake, such as passing one of these to `insertOne()` and forgetting to specify a `codec` option.

And I agree that Unserializable is unnecessary, since these objects are instantiated using a codec.

Reply:
Looking at implementing `Serializable`, this would indeed profit from us allowing `bsonSerialize` returning more than just arrays and `stdClass` instances (see PHPC-1000). I've created PHPLIB-1214 to track this separately so we can close this PR out while I work on the PHPC changes.